File: contrib/plugins/inflation_calc/inflation_calc.rb
#--
##   Riassence Framework
 #   Copyright 2009 Riassence Inc.
 #   http://riassence.com/
 #
 #   You should have received a copy of the GNU General Public License along
 #   with this software package. If not, contact licensing@riassence.com
 ##
 #++

# = Inflation Calculator
# This plugin demonstrates using a separate library file for
# calculating values and how to initialize values and user interfaces
# without YAML.
#
$LOAD_PATH << PluginManager.curr_plugin_path
require 'lib/inflation_calc_responder'
class InflationCalc < Plugin
  
  # See lib/inflation_calc_responder, it contains the calculate method.
  include InflationCalcResponder
  
  # Initialize session structure
  def init_ses( msg )
    
    # Create values for the fields in the user interface
    msg.session[:inflation] = {
      :percent =>  HValue.new(msg, 7 ),
      :amount  =>  HValue.new(msg, 1000 ),
      :years   =>  HValue.new(msg, 10 ),
      :result_future =>  HValue.new(msg, [] ),
      :result_past   =>  HValue.new(msg, [] )
    }
    
    # Bind three of them (input data) to the calculate method as the responder.
    ses = msg.session[:inflation]
    [ :percent, :amount, :years ].each do |val|
      ses[val].bind('inflation','calculate')
    end
    
  end
  
  # Initializes the user interface "manually"
  def init_ui( msg )
    
    # See init_ses
    ses = msg.session[:inflation]
    
    # Dependency packages
    include_js( msg, ['controls','default_theme'] )
    
    # Loads the SimpleTable class
    msg.reply read_js_once( msg, 'simple_table' )
    
    # The user interface code as js
    msg.reply read_js_once( msg, 'inflation_calc' )
    
    # Initial calculation
    calculate( msg )
    
    # Constructs an instance of the user interface
    msg.reply "InflationCalc.nu(#{values_js(ses)});"
    
  end
end
InflationCalc.new.register('inflation')

File: contrib/plugins/inflation_calc/js/inflation_calc.js
/*   Riassence Framework
 *   Copyright 2009 Riassence Inc.
 *   http://riassence.com/
 *
 *   You should have received a copy of the GNU General Public License along
 *   with this software package. If not, contact licensing@riassence.com
 */


/*** = Description
  ** A self-contained application class that creates the user interface.
  ***/
InflationCalc = HApplication.extend({
  
/** Resets the values to the defaults.
  **/
  reset: function(){
    var values = this.values;
    values.amount.set(1000);
    values.years.set(10);
    values.percent.set(7);
  },
  
/** Uses +values+ as the constructor parameter, it's a hash
  * that contains the references to the values used in the application.
  * See inflation_calc.rb
  **/
  constructor: function(values){
    this.values = values;
    this.base();
    
    // Extends the HWindow component in place to contain the gui elements.
    this.view = HWindow.extend({
      drawSubviews: function(){
        // These are constructed inside the window.
        
        // Label for the value field
        HStringView.nu( HRect.nu(8,22,150,45),this,{value:'<b>Current value:</b>'});
        
        // The value field itself
        HTextControl.nu(HRect.nu(150,20,250,40),this,{valueObj:this.app.values.amount,events:{textEnter:true}});
        
        // The 'Reset' button, calls the reset method of the application upon click.
        HButton.extend({
          click: function(){
            this.app.reset();
          }
        }).nu(HRect.nu(258,18,330,42),this,{events:{click:true},label:'Reset'});
        
        // Label for the inflation rate field
        HStringView.nu( HRect.nu(8,52,150,75),this,{value:'<b>Inflation rate:</b>'});
        
        // The inflation rate field itself
        HTextControl.nu(HRect.nu(150,50,190,70),this,{valueObj:this.app.values.percent,events:{textEnter:true}});
        
        // Percent as the suffix for the field
        HStringView.nu( HRect.nu(190,52,210,75),this,{value:'%'});
        
        // A slider to control the rate
        HSlider.extend({
          setValue: function(value){
            if(isNaN(value)){return;}
            this.base(Math.round(value*10)/10); // makes decimals with just one digit after the dot
          }
        }).nu(HRect.nu(200,50,330,70),this,{valueObj:this.app.values.percent,minValue:-20,maxValue:20});
        
        // Label for the years field
        HStringView.nu( HRect.nu(8,82,150,105),this,{value:'<b>Number of years:</b>'});
        
        // The years field
        HTextControl.nu(HRect.nu(150,80,190,100),this,{valueObj:this.app.values.years,events:{textEnter:true}});
        
        // A slider to control the years
        HSlider.extend({
          setValue: function(value){
            if(isNaN(value)){return;}
            this.base(Math.round(value)); // makes integers
          }
        }).nu(HRect.nu(200,80,330,100),this,{valueObj:this.app.values.years,minValue:1,maxValue:100});
        
        // Label for the future values table
        HStringView.nu( HRect.nu(8,120,158,140),this,{value:'<b>Future value:</b>'});
        
        // Table for the future values
        SimpleTable.nu(HRect.nu( 8,140,158,300),this,{valueObj:this.app.values.result_future});
        
        // Label for the past values table
        HStringView.nu( HRect.nu(180,120,330,140),this,{value:'<b>Past value:</b>'});
        
        // Table for the past values
        SimpleTable.nu(HRect.nu(180,140,330,300),this,{valueObj:this.app.values.result_past  });
      }
      
      // Constructor arguments for the HWindow extended above
    }).nu(HRect.nu(40,40,386,508),this,{
      label: 'Inflation calculator',
      zoomButton: true,
      collapseButton: true,
      closeButton: true,
      minSize: [346,380],
      maxSize: [346,3072]
    });
  }
});

